### PR TITLE
ITX5OTHER-739 Implement user login/registration 

### PR DESCRIPTION
**Pull request by OsipenkoDenis**
Tiket's requirements:
1. Login/registration forms,
2. Interceptor to catch and use token in each request.

### CODE REVIEW

File: src/app/app.component.ts [hunk 1]
```diff
@@ -1,9 +1,19 @@
...
 @Component({
   selector: 'app-root',
   templateUrl: './app.component.html',
   styleUrls: ['./app.component.scss']
 })
-export class AppComponent {
+export class AppComponent implements OnInit{
+
+  constructor(private auth: AuthService) {
+  }
```
Comment (romaaxa) [inline src/app/app.component.ts:12]: add space between methods and please use code formatter to get rid of these issues

File: src/app/app.component.ts [hunk 2]
```diff
@@ -1,9 +1,19 @@
...
+  ngOnInit(){
+  const token = localStorage.getItem('auth-token')
+    if(token !== null){
```
Comment (romaaxa) [inline src/app/app.component.ts:15]: what if token is undefined? can be just `if (token)`

File: src/app/models/product.interface.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export interface Product {
```
Comment (romaaxa) [inline src/app/models/product.interface.ts:1]: not a nitpick but its better to name interfaces as `product.model.ts`,types .type.ts and enums `.enum.ts`

File: src/app/pages/home-page/home-page-presentation.component.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+})
+export class HomePagePresentationComponent {
+
+  @Input()
+  isAuthenticated?:boolean = false
+
+  @Output()
+  logOutEvent: EventEmitter<Event> = new EventEmitter<Event>();
+
+  logout(event: Event) {
```
Comment (romaaxa) [inline src/app/pages/home-page/home-page-presentation.component.ts:17]: add return type

File: src/app/pages/home-page/home-page-presentation.component.ts [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+  templateUrl: './home-page.component.html',
+  styleUrls: ['./home-page.component.scss']
+})
+export class HomePagePresentationComponent {
+
+  @Input()
+  isAuthenticated?:boolean = false
+
+  @Output()
+  logOutEvent: EventEmitter<Event> = new EventEmitter<Event>();
```
Comment (romaaxa) [inline src/app/pages/home-page/home-page-presentation.component.ts:15]: do not name like logOutEvent, ts - logout and in html method would be named like `onLogout`

File: src/app/pages/home-page/home-page-smart.component.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  styleUrls: ['home-page.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+
+export class HomePageSmartComponent {
+  constructor(public auth: AuthService,
+              private router: Router) {
+  }
+
+  logout(event: Event){
```
Comment (romaaxa) [inline src/app/pages/home-page/home-page-smart.component.ts:20]: add return type please

File: src/app/pages/home-page/home-page.component.html [hunk 7]
```diff
@@ -0,0 +1,31 @@
+<nav class='home-page-nav navbar bg-dark'>
```
Comment (romaaxa) [inline src/app/pages/home-page/home-page.component.html:1]: please wrap or use -wrapper, -container classes

File: src/app/pages/login-page/login-page-presentation.component.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+@Component({
+  selector: 'login-page-presentation',
+  templateUrl:'./login-page.component.html',
+  styleUrls:['./login-page.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush
+})
+
+export class LoginPagePresentationComponent implements OnInit{
+  form!:FormGroup
```
Comment (romaaxa) [inline src/app/pages/login-page/login-page-presentation.component.ts:13]: follow the sequence: private/readonly variables, viewchild, input, output, public varaibles, constructor, methods`

File: src/app/pages/login-page/login-page-presentation.component.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  userLoginValue:EventEmitter<User> = new EventEmitter<User>();
+
+  ngOnInit(): void {
+    this.form = new FormGroup({
+      "username": new FormControl( '', [Validators.required]),
+      "password": new FormControl( '', [Validators.required, Validators.minLength(8)])
+    });
+  }
+
+  onLogin() {
```
Comment (romaaxa) [inline src/app/pages/login-page/login-page-presentation.component.ts:26]: add return type

File: src/app/pages/login-page/login-page-presentation.component.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    if(this.form.invalid){
+      return;
+    }
+    this.isLogin = true;
```
Comment (romaaxa) [inline src/app/pages/login-page/login-page-presentation.component.ts:30]: whats the purpose of this variable? naming should be more specific

File: src/app/pages/login-page/login-page-smart.component.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+import { Router } from '@angular/router';
+import { Subscription } from 'rxjs';
+import { User } from '../../models/user.interface';
+import { AuthService } from '../services/auth.service';
+
+@Component({
+  selector: 'login-page-smart',
+  template: `
+    <login-page-presentation
+            (userLoginValue)='onLogin($event)'></login-page-presentation>`,
```
Comment (romaaxa) [inline src/app/pages/login-page/login-page-smart.component.ts:11]: closing tag should be on new line here

File: src/app/pages/login-page/login-page-smart.component.ts [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class LoginPageSmartComponent implements OnDestroy {
+  authSubscription!: Subscription;
+
+  constructor(private auth: AuthService,
+              private router: Router) {
+  }
+
+  ngOnDestroy() {
```
Comment (romaaxa) [inline src/app/pages/login-page/login-page-smart.component.ts:22]: add return type

File: src/app/pages/login-page/login-page-smart.component.ts [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    if (this.authSubscription) {
+      this.authSubscription.unsubscribe();
+    }
+  }
+
+  onLogin(userValue: User): void {
```
Comment (romaaxa) [inline src/app/pages/login-page/login-page-smart.component.ts:28]: `onLogin(userInfo...`

File: src/app/pages/register-page/register-page-presentation.component.ts [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+  selector: 'register-page-presentation',
+  templateUrl: './register-page.component.html',
+  styleUrls: ['./register-page.component.scss']
+})
+export class RegisterPagePresentationComponent implements OnInit {
+  form!:FormGroup
+  isRegistered:Boolean = false;
+
+  @Output()
+  userRegistrationValue:EventEmitter<User> = new EventEmitter<User>();
```
Comment (romaaxa) [inline src/app/pages/register-page/register-page-presentation.component.ts:15]: follow sequence

File: src/app/pages/register-page/register-page-presentation.component.ts [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+  constructor() { }
```
Comment (romaaxa) [inline src/app/pages/register-page/register-page-presentation.component.ts:17]: redundant

File: src/app/pages/register-page/register-page-presentation.component.ts [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+  ngOnInit(): void {
+    this.form = new FormGroup({
+      "username": new FormControl( '', [Validators.required]),
+      "password": new FormControl( '', [Validators.required, Validators.minLength(8)])
+    });
+  }
+
+  onRegister() {
```
Comment (romaaxa) [inline src/app/pages/register-page/register-page-presentation.component.ts:26]: add return type

File: src/app/pages/services/auth.service.ts [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+import { User } from '../../models/user.interface';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class AuthService {
+  private token: string = '';
+
+  constructor(private http:HttpClient) { }
+
```
Comment (romaaxa) [inline src/app/pages/services/auth.service.ts:14]: this service should be in the auth module

File: src/app/pages/services/auth.service.ts [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+          }
+        )
+      )
+  }
+
+  setToken(token:string){
+    this.token = token
+  }
+
+  getToken(){
```
Comment (romaaxa) [inline src/app/pages/services/auth.service.ts:34]: add return types here everywhere

File: src/app/models/review.interface.ts [hunk 19]
```diff
@@ -0,0 +1,4 @@
+export interface Review {
```
Comment (romaaxa) [inline src/app/models/review.interface.ts:1]: also if model is used only in one module then there is no need to move it to common module

File: src/app/app.component.ts (hunk 1 above)
Comment (OsipenkoDenis) [reply src/app/app.component.ts:12]: Resolved

File: src/app/app.component.ts (hunk 2 above)
Comment (OsipenkoDenis) [reply src/app/app.component.ts:15]: Resolved

File: src/app/models/product.interface.ts (hunk 3 above)
Comment (OsipenkoDenis) [reply src/app/models/product.interface.ts:1]: Resolved

File: src/app/pages/home-page/home-page-presentation.component.ts (hunk 4 above)
Comment (OsipenkoDenis) [reply src/app/pages/home-page/home-page-presentation.component.ts:17]: Resolved

File: src/app/pages/home-page/home-page-presentation.component.ts (hunk 5 above)
Comment (OsipenkoDenis) [reply src/app/pages/home-page/home-page-presentation.component.ts:15]: Resolved

File: src/app/pages/home-page/home-page-smart.component.ts (hunk 6 above)
Comment (OsipenkoDenis) [reply src/app/pages/home-page/home-page-smart.component.ts:20]: Resolved

File: src/app/pages/home-page/home-page.component.html (hunk 7 above)
Comment (OsipenkoDenis) [reply src/app/pages/home-page/home-page.component.html:1]: Resolved

File: src/app/pages/login-page/login-page-smart.component.ts (hunk 12 above)
Comment (OsipenkoDenis) [reply src/app/pages/login-page/login-page-smart.component.ts:22]: Resolved

File: src/app/pages/register-page/register-page-presentation.component.ts (hunk 15 above)
Comment (OsipenkoDenis) [reply src/app/pages/register-page/register-page-presentation.component.ts:17]: Resolved

File: src/app/pages/services/auth.service.ts (hunk 18 above)
Comment (OsipenkoDenis) [reply src/app/pages/services/auth.service.ts:34]: Resolved

File: src/app/pages/services/auth.service.ts (hunk 17 above)
Comment (OsipenkoDenis) [reply src/app/pages/services/auth.service.ts:14]: Resolved

File: src/app/models/review.interface.ts (hunk 19 above)
Comment (OsipenkoDenis) [reply src/app/models/review.interface.ts:1]: Resolved

File: src/app/pages/login-page/login-page-presentation.component.ts (hunk 10 above)
Comment (OsipenkoDenis) [reply src/app/pages/login-page/login-page-presentation.component.ts:30]: This variable uses to block button "Login" after submitting login and password(to get rid new additional requests).

Comment (romaaxa) [reply src/app/pages/login-page/login-page-presentation.component.ts:31]: then `isLogged`?

Comment (OsipenkoDenis) [reply src/app/pages/login-page/login-page-presentation.component.ts:31]: Yes, that's better then mine.

**Pull request merged**